MMM-MBTA.js

```diff
@@ -277,86 +404,199 @@ Module.register("MMM-MBTA", {
+        var alertPromises = [];
+        var self = this;
+        for (i = 0; i < alertIds.length; i++) {
+            var alertPromise = $.Deferred();
+            alertPromises.push(alertPromise);
+
+            var alertId = alertIds[i].id
+
+            var alertUrl = this.detailsUrl("alerts", alertId);
+            var alertRequest = new XMLHttpRequest();
+            alertRequest.open("GET", alertUrl, true);
+
+            alertRequest.onreadystatechange = function() {
+                if (this.readyState === 4) {
+                    if (this.status === 200) {
+                        alertPromise.resolve(JSON.parse(this.response));
+                    } else if (this.status === 404) {
+                        alertPromise.resolve("Unavailable");
+                    }
+                }
+            };
+            alertRequest.send();
         }
 
+        $.when(...alertPromises).then(function(...alertsParse) {
+            self.processData(data, updateDomAfter, pred, routeParse, tripParse, alertsParse, rawData);
+            promise.resolve();
+        });
+    },
+
+    detailsUrl: function(detail,id) {
+        var url = this.config.baseUrl;
+        url += detail + '/';
+        url += id;
+        url += "?api_key=" + this.config.apikey;
+
+        return url
+    },
+
+    loopData: function(data, updateDomAfter) {
+        this.stationData = [ ]; // clear all data.
+        var rawData = [ ];
+        var promises = [];
+        if (data.data.length === 0) {
+            this.loaded = true;
+            this.updateDom();
+        } else {
+            for (let pred = 0; pred < data["data"].length; pred++) {
+                routeId = data.data[pred].relationships.route.data.id;
+                tripId = data.data[pred].relationships.trip.data.id;
+                alertIds = data.data[pred].relationships.alerts.data;
+debugger
+                promises.push(this.fetchRoute(data, updateDomAfter, pred, routeId, tripId, alertIds, rawData));
+            }
+
+            var that = this;
+            $.when(...promises).then(function() {
+                that.cleanData(data, rawData, updateDomAfter);
+            });
+        }
+    },
+    
+    // updateDomAfter: immediatelly call updateDom() if true
+    processData: function(data, updateDomAfter, pred, routeParse, tripParse, alertsParse, rawData) {
+        /* Each element in this array is an entry on our displayed table:
+        {"routeType": string,
+         "routeId": string,
+         "tripSign": string,
+         "directionId": string,
+         "alerts": array,
+         "preDt": int,
+         "preArr": int,
+         "preETA": int
+        } */
+
+        alertsArray = [];
+        for (i = 0; i < alertsParse.length; i++) {
+            if (alertsParse[i].data) {
+                alert = alertsParse[i].data.attributes.header;
+                alertsArray.push(alert);
+            }
+        };
```

Comment:
extra semi colon

Reply:
Actually if you want this to look cleaner you can do a for...of loop:
```js
for (let alert of alertsParse) {
   if (alert.data) {
        alertsArray.push(alert.data.attributes.header);
    }
}
```

Reply:
Optional ofc, but nice learning point.

Reply:
I'm all about cleaner code. I'll change it.